src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Lower.java

```diff
@@ -3518,16 +3518,16 @@ private void visitIterableForeachLoop(JCEnhancedForLoop tree) {
                                               syms.iterableType.tsym);
             if (iterableType.getTypeArguments().nonEmpty())
                 iteratorTarget = types.erasure(iterableType.getTypeArguments().head);
-            Type eType = types.skipTypeVars(tree.expr.type, false);
-            tree.expr.type = types.erasure(eType);
-            if (eType.isCompound())
+            tree.expr.type = types.erasure(types.skipTypeVars(tree.expr.type, false));
+            if (types.asSuper(tree.expr.type, syms.iterableType.tsym) == null)
                 tree.expr = make.TypeCast(types.erasure(iterableType), tree.expr);
             Symbol iterator = lookupMethod(tree.expr.pos(),
                                            names.iterator,
-                                           eType,
+                                           tree.expr.type,
                                            List.nil());
+            Assert.check(types.isSameType(types.erasure(types.asSuper(iterator.type.getReturnType(), syms.iteratorType.tsym)), types.erasure(syms.iteratorType)));
             VarSymbol itvar = new VarSymbol(SYNTHETIC, names.fromString("i" + target.syntheticNameChar()),
-                                            types.erasure(types.asSuper(iterator.type.getReturnType(), syms.iteratorType.tsym)),
+                                            types.erasure(syms.iteratorType),
                                             currentMethodSym);
```

Comment:
The original code seems to be a convoluted way of declaring an Iterator. Replace with a simpler more direct expression